packages/navie/src/services/anthropic-completion-service.ts

```diff
@@ -132,39 +157,59 @@ export default class AnthropicCompletionService implements CompletionService {
   }
 
   async *complete(messages: readonly Message[], options?: { temperature?: number }): Completion {
-    try {
-      const model = this.buildModel(options);
-
-      const sentMessages = mergeSystemMessages(messages);
-      for (const message of sentMessages) this.trajectory.logSentMessage(message);
-
-      const response = await model.stream(sentMessages.map(convertToMessage));
+    const usage = new Usage(COST_PER_M_TOKEN[this.modelName]);
+    const model = this.buildModel(options);
+    const sentMessages = mergeSystemMessages(messages);
+    for (const message of sentMessages) this.trajectory.logSentMessage(message);
 
-      const usage = new Usage(COST_PER_M_TOKEN[this.modelName]);
+    const maxAttempts = CompletionRetries;
+    for (let attempt = 0; attempt < maxAttempts; attempt += 1) {
+      try {
+        const tokens = new Array<string>();
+
+        // eslint-disable-next-line no-await-in-loop
+        const response = await model.stream(sentMessages.map(convertToMessage));
+
+        // eslint-disable-next-line @typescript-eslint/naming-convention, no-await-in-loop
+        for await (const { content, usage_metadata } of response) {
+          yield content.toString();
+          tokens.push(content.toString());
+          if (usage_metadata) {
+            usage.promptTokens += usage_metadata.input_tokens;
+            usage.completionTokens += usage_metadata.output_tokens;
+          }
+        }
 
-      const tokens = new Array<string>();
-      // eslint-disable-next-line @typescript-eslint/naming-convention
-      for await (const { content, usage_metadata } of response) {
-        yield content.toString();
-        tokens.push(content.toString());
-        if (usage_metadata) {
-          usage.promptTokens += usage_metadata.input_tokens;
-          usage.completionTokens += usage_metadata.output_tokens;
+        this.trajectory.logReceivedMessage({
+          role: 'assistant',
+          content: tokens.join(''),
+        });
+
+        break;
+      } catch (cause) {
+        if (attempt < maxAttempts - 1) {
+          const sseError = getSseError(cause);
```

Comment:
You have to throw if you've already yielded any tokens. Compare with https://github.com/getappmap/appmap-js/pull/2000 (which is the same issues mod specifics of the error); btw, I wonder if we should DRY this up somehow, we're basically maintaining two separate implementation of almost the same thing.

Reply:
Good point. I've changed it to bail if any tokens have been written. I haven't made an effort to DRY this up as of yet, I'd like to see it working first.

Reply:
It definitely seemed to be working in the benchmark runs. Don't make any risky changes that would break it again :-)